java/vector/src/main/java/org/apache/arrow/vector/BitVector.java

```diff
@@ -276,36 +276,55 @@ public void transferTo(BitVector target) {
 
   public void splitAndTransferTo(int startIndex, int length, BitVector target) {
     assert startIndex + length <= valueCount;
-    int firstByte = getByteIndex(startIndex);
-    int byteSize = getSizeFromCount(length);
+    int firstByteSource = getByteIndex(startIndex);
+    int lastByteSource = getByteIndex(valueCount-1);
+    int byteSizeTarget = getSizeFromCount(length);
     int offset = startIndex % 8;
+
     if (length > 0) {
       if (offset == 0) {
         target.clear();
         // slice
         if (target.data != null) {
           target.data.release();
         }
-        target.data = data.slice(firstByte, byteSize);
+        target.data = data.slice(firstByteSource, byteSizeTarget);
         target.data.retain(1);
-      } else {
+      }
+      else {
         // Copy data
         // When the first bit starts from the middle of a byte (offset != 0), copy data from src BitVector.
         // Each byte in the target is composed by a part in i-th byte, another part in (i+1)-th byte.
-        // The last byte copied to target is a bit tricky :
-        //   1) if length requires partly byte (length % 8 !=0), copy the remaining bits only.
-        //   2) otherwise, copy the last byte in the same way as to the prior bytes.
+
         target.clear();
-        target.allocateNew(length);
+        target.allocateNew(byteSizeTarget * 8);
+        int destPos = 0;
+        int sourcePos = startIndex;
         // TODO maybe do this one word at a time, rather than byte?
-        for (int i = 0; i < byteSize - 1; i++) {
-          target.data.setByte(i, (((this.data.getByte(firstByte + i) & 0xFF) >>> offset) + (this.data.getByte(firstByte + i + 1) << (8 - offset))));
+
+        int counter = firstByteSource;
+
+        for (int i = 0; i < byteSizeTarget - 1; i++) {
+          target.data.setByte(i,
+                  (((this.data.getByte(firstByteSource + i) & 0xFF) >>> offset) + (this.data.getByte(firstByteSource + i + 1) << (8 - offset))));
```

Comment:
Maybe refactor this line to a function?

Reply:
For this kind of thing, I'm less excited about extra functions. If extra function, should probably be private static. Also best to final the arguments references.

Reply:
+1